proposals/22-03-value-in-the-structured-data.md

```diff
@@ -0,0 +1,569 @@
+Use the status field to record the results of the latest review:
+- implementable: The default for this repo. If the proposal is merged, you can start working on it.
+- deferred: The proposal may be accepted in the future, but it has been shelved for the time being.
+A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
+- withdrawn: The author changed their mind and no longer wants to pursue the proposal.
+A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
+- rejected: This proposal should not be implemented.
+- replaced: If you submit a new proposal that supersedes an older one,
+update the older one's status to "replaced by <link>".
+-->
+
+## Summary
+
+This proposal decides the interfaces to change values in the structured data (like json,yaml) inside a Kubernetes objects' field value and implements changing function target a few formats (mainly json).
+
+
+## Motivation
+
+<!--
+If this proposal is an expansion of an existing GitHub issue, link to it here.
+-->
+
+kustomize can apply structured edits to Kubernetes objects defined in yaml files.\
+Sometimes structured multi-line or long single line string (ex. json,yaml, and other structured format data) is injected in Kubernetes objects' string literal field. From the kustomize perspective, these "structured" multiline strings are just arbitrary unstructured strings.\
+So, kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's string literal field. This function is expected behavior, but kustomize will be very helpful if it can change the value of structured data like json and yaml substrings in a string literal.\
+This proposal allows the user to identify strings literals containing JSON/YAML data so that Kustomize can make structured edits to the data they contain. This allows the requested functionality to be added without violating Kustomize's core principle of supporting structured edits exclusively.\
+
+For example, kustomize can't change the value `"REPLACE_TARGET_HOSTNAME"` in this yaml file straightforwardly.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: target-configmap
+data:
+  config.json: |-
+    {"config": {
+      "id": "42",
+      "hostname": "REPLACE_TARGET_HOSTNAME"
+    }}
+```
+
+This function has been wanted for a long time.
+- https://github.com/kubernetes-sigs/kustomize/issues/680
+- https://github.com/kubernetes-sigs/kustomize/issues/3787
+- https://github.com/kubernetes-sigs/kustomize/issues/4517
+
+This function can be an alternative deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/) function in many use cases.
+
+**Goals:**
+<!--
+List the specific goals of the proposal. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. Provide the way to update values in the structured data like kubernetes objects with JSON/YAML formats.
```

Comment:
I have been observing this for some time because I was bitten by the problem this proposal is trying to address. Unfortunately, I believe, this cannot be solved 100% to the user's expectation.

Various operators put various data in their config maps. My biggest question to this proposal is, what is going to happen when a config map has something called JSON but that JSON is actually a template?

```yaml
# base/kustomization.yaml
configMapGenerator:
- name: demo
  literals:
  - config.json: |-
      {
        "config": {
          "loglevel": debug,
          "parameter": {
            "foo": {{ I am some golang template }}
          }
        }
      }
```

Reply:
Is this proposed idea deprecated by or duplicative of `transformers:` plugins? Maybe more clear that all "opaque" data handling is done in plugins, and not possibly inline? There could be a plugin per transform type like jsonStrategicMerge, json6902, [jq](https://stedolan.github.io/jq/), [yq](https://stedolan.github.io/jq/), etc...

I have templates in config/secret data as well like mentioned by @radekg. The base config uses [mustache](https://mustache.github.io/), and in an overlay I'm applying a custom transformer following the example at https://github.com/badjware/kustomize-plugins/tree/61701852ee3e9873ca9ceed68c0936c28c06d7d5/badjware/v1/placeholdertransformer which pulls in data externally based on my plugin's configuration (from Google Secret Manager). Although I do have to specify `--enable-alpha-plugins` and `KUSTOMIZE_PLUGIN_HOME`, it seems to work great.

Reply:
@radekg I think you ask an important question.

In my principle uses for k-vars today, the base is generated from helm and I have injected uniquely identifiable marker strings into these rendered manifests specifically so that they can be replaced in kustomized overlays using vars.

Replacing strings that have been carefully injected into strangely formatted data is not an edge case.

I have provided the "structure" that is needed for replacement by carefully choosing these marker strings.